File: 52.n-皇后-ii.cpp
/*
 * @lc app=leetcode.cn id=52 lang=cpp
 *
 * [52] N 皇后 II
 */

// @lc code=start
class Solution {
public:
    int res = 0;
    vector<bool> arr1, arr2, cols; 
    int totalNQueens(int n) {
        // 用哈希表分别检查 r+c & r-c 是否出现过
        // 因为45`，r-c的值就是相等的
        // 135`时上r+c=n
        cols.resize(n, false);
        arr1.resize(2 * n, false);
        // 在左上角处会出现 -(n-1) 所以需要加上保护值 n-1
        arr2.resize(2 * n, false);
        backtrack(n, 0);
        return res;
    }

    void backtrack(int n, int row) {
        if(row == n) {
            res++;
            return;
        }

        for(int col = 0; col < n; col++) {
            if(!cols[col] && !arr1[row + col] && !arr2[row - col + n]) {
                arr2[row - col + n] = true;
                arr1[row + col] = true;
                cols[col] = true;
                backtrack(n, row + 1);
                cols[col] = false;
                arr1[row + col] = false;
                arr2[row - col + n] = false;
            }
        }
    }
};
// @lc code=end

File: 52.n-皇后-ii.py
#
# @lc app=leetcode.cn id=52 lang=python3
#
# [52] N 皇后 II
#

# @lc code=start
class Solution:
    def totalNQueens(self, n: int) -> int:
        cols = [False] * n
        arr1 = [False] * (2 * n - 1)
        arr2 = [False] * (2 * n - 1)
        self.res = 0
        board = [['.' for _ in range(n)] for _ in range(n)]
        def backtrack(row: int):
            if row == n:
                self.res += 1
                return
            
            for col in range(n):
                if cols[col] or arr1[row + col] or arr2[row - col + n - 1]:
                    continue
                
                arr1[row + col] = True
                arr2[row - col + n - 1] = True
                cols[col] = True
                board[row][col] = 'Q'
                backtrack(row + 1)
                arr1[row + col] = False
                arr2[row - col + n - 1] = False
                cols[col] = False
                board[row][col] = '.'
        backtrack(0)
        return self.res

            
                

        
# @lc code=end

